src/durmod/kaldi-durmod.cc

```diff
@@ -0,0 +1,466 @@
+  // that the probabilities do not decline too rapidly.
+
+  int32 actual_duration_id = phone_duration - 1;  // in case max duration was
+                                                  // infinity (i.e. we
+                                                  // had infinitely many nodes
+                                                  // at the output of the
+                                                  // network)
+  BaseFloat logprob = output(0, duration_id);
+
+  // make sure the distribution over all durations (1 to inf) sums to 1
+  BaseFloat alpha = Exp(-1.0f / model_.MaxDuration());
+  BaseFloat probability_normalization_sum = alpha / (1 - alpha);
+  if (phone_duration >= model_.MaxDuration())
+    logprob += Log(pow(alpha, actual_duration_id - duration_id + 1) /
+                                                 probability_normalization_sum);
+  return logprob;
+}
+
+PhoneDurationModelDeterministicFst::PhoneDurationModelDeterministicFst(
+                                      int32 num_phones,
+                                      const PhoneDurationModel &model,
+                                      NnetPhoneDurationScoreComputer *scorer):
+                            context_size_(model.FullContextSize()),
+                            right_context_(model.RightContext()),
+                            num_phones_(num_phones),
+                            scorer_(*scorer) {
+  // In the beginning we have a Null left context.
+  // Null means that the phone-identity and duration are both zero --> olabel=0
+  std::vector<int32> start_context(model.LeftContext(), 0);
+  state_to_context_.push_back(start_context);
+  context_to_state_[start_context] = 0;
+  start_state_ = 0;
+}
+
+fst::StdArc::Weight PhoneDurationModelDeterministicFst::Final(StateId state) {
+  KALDI_ASSERT(state < static_cast<StateId>(state_to_context_.size()));
+  std::vector<Label> seq = state_to_context_[state];
+  BaseFloat logprob = 0;
+  // handle the phones that have not been considered yet:
+  for (int i = 0; i < right_context_; i++) {
+    seq.push_back(0);  // add Null right context
+    if (seq.size() >= context_size_) {
+      if (seq.size() > context_size_)
+        seq.erase(seq.begin());
+      logprob += GetLogProb(seq);
+    }
+  }
+  return Weight(-logprob);
+}
+
+bool PhoneDurationModelDeterministicFst::GetArc(StateId s,
+                                                Label ilabel,
+                                                fst::StdArc *oarc) {
+  // The state ids increment with each state we encounter.
+  // if the assert fails, then we are trying to access
+  // unseen states that are not immediately traversable.
+  KALDI_ASSERT(static_cast<size_t>(s) < state_to_context_.size());
+  std::vector<Label> seq = state_to_context_[s];
+
+  oarc->ilabel = ilabel;
+  oarc->olabel = ilabel;
+  if (ilabel == 0) {  // if this is an epsilon arc, move on
+    oarc->nextstate = s;
```

Comment:
is this code ever reached?  IIRC, the interface for DeterministicOnDemandFst doesn't allow requests for epsilon arcs.